tests/compile-fail/validation_lock_confusion.rs

```diff
@@ -0,0 +1,24 @@
+// Make sure validation can handle many overlapping shared borrows for different parts of a data structure
+#![allow(unused_variables)]
+use std::cell::RefCell;
+
+fn evil(x: *mut i32) {
+    unsafe { *x = 0; } //~ ERROR: in conflict with lock WriteLock
+}
+
+fn test(r: &mut RefCell<i32>) {
+    let x = &*r; // releasing write lock, first suspension recorded
```

Comment:
If you want you can try using `ui` tests together with `MIRI_LOG=rust_miri::interpret::validate=trace` to be able to ensure that the trace does exactly what you want.

But I guess it would fail with an interpretation error if anything went wrong...

Reply:
The test fails prior to the PR, so I think compile-fail is enough here.